cmd/kubectl_token.go

```diff
@@ -510,58 +532,84 @@ func getAuthProviders(server string) (map[string]string, error) {
 		return nil, err
 	}
 
-	data := map[string]interface{}{}
-	err = json.Unmarshal(response, &data)
-	if err != nil {
-		return nil, err
+	if !gjson.ValidBytes(response) {
+		return nil, errors.New("invalid JSON input")
 	}
 
-	providers := map[string]string{}
-	i := 0
-	for _, value := range convert.ToMapSlice(data["data"]) {
-		provider := convert.ToString(value["type"])
-		if provider != "" && supportedAuthProviders[provider] {
-			providers[fmt.Sprintf("%v", i)] = provider
-			i++
+	data := gjson.GetBytes(response, "data").Array()
+
+	supportedProviders := []TypedProvider{}
+	for _, provider := range data {
+		providerType := provider.Get("type").String()
+
+		if providerType != "" && supportedAuthProviders[providerType] {
+			var typedProvider TypedProvider
+
+			switch providerType {
+			case "azureADProvider":
+				typedProvider = &apiv3.AzureADProvider{}
+			case "localProvider":
+				typedProvider = &apiv3.LocalProvider{}
+			default:
+				typedProvider = &apiv3.AuthProvider{}
+			}
+
+			err = json.Unmarshal([]byte(provider.Raw), typedProvider)
+			if err != nil {
+				return nil, err
```

Comment:
We could maybe wrap this with something that says "attempting to decode the auth provider of type %s", otherwise these sorts of errors are quite generic (and thus opaque).